src/account/accountsReducer.js

```diff
@@ -110,6 +112,34 @@ export default (state: AccountsState = initialState, action: Action): AccountsSt
     case ACCOUNT_REMOVE:
       return accountRemove(state, action);
 
+    case EVENT: {
+      const { event } = action;
+      switch (event.type) {
+        case 'restart': {
+          const { zulip_feature_level, zulip_version } = event;
+          if (zulip_feature_level !== undefined && zulip_version !== undefined) {
+            // TODO (?): Detect if these are different from the values
+            // we had before, so we know it's an upgrade, not just a
+            // restart. Then, implement logic like "be sure to refetch
+            // from scratch within N hours" (to avoid thundering
+            // herding the server).
+            return [
+              {
+                ...state[0],
+                zulipVersion: new ZulipVersion(zulip_version),
+                zulipFeatureLevel: zulip_feature_level ?? 0,
+              },
+              ...state.slice(1),
+            ];
+          } else {
+            return state;
+          }
```

Comment:
It might be simpler to do:

```javascript
return [
  {
    ...state[0],
    zulipVersion: zulip_version ? new ZulipVersion(zulip_version) : state[0].zulipVersion,
    zulipFeatureLevel: zulip_feature_level ? zulip_feature_level : state[0].zulipFeatureLevel,
  },
  ...state.slice(1),
];
```

I don't have a strong feeling that this would be better, just something to think about.

Reply:
Yeah, possibly. One feature of the code in this revision is that `state` won't be replaced in cases where it doesn't meaningfully change (and so components that subscribe to it using react-redux won't rerender, etc.). But that optimization isn't really important at all because these events will be rare.

I think we could still get that feature while using a conditional operator, and so only having one `return` statement which I kind of like; I'll try that?